README.md
![License](https://img.shields.io/github/license/zaphyr-org/encrypt?style=for-the-badge)
![Version](https://img.shields.io/packagist/v/zaphyr-org/encrypt?style=for-the-badge)
![Downloads](https://img.shields.io/packagist/dt/zaphyr-org/encrypt?style=for-the-badge)
![Stars](https://img.shields.io/github/stars/zaphyr-org/encrypt?style=for-the-badge)
![Issues](https://img.shields.io/github/issues/zaphyr-org/encrypt?style=for-the-badge)

# ENCRYPT

A convenient interface for encrypting and decrypting text via OpenSSL using AES-128 and AES-256 encryption.

## System Requirements

You need `PHP => 8.1.0` but the latest stable version of PHP is recommended.

## Installation

It's recommended that you use [Composer](https://getcomposer.org/) to install the encrypt repository.

```console
composer require zaphyr-org/encrypt
```

## Documentation

You can read the full documentation [here](https://zaphyr.org/docs/2.x/repositories/encrypt).

## Issues

Bug reports and feature requests can be submitted on the [GitHub Issue Tracker](https://github.com/zaphyr-org/encrypt/issues).

## Security

If you discover security related issues, please email security@zaphyr.org **instead of using
the issue tracker!**

## Contributing

Please see [CONTRIBUTING](https://zaphyr.org/contributions) for details.

## Code of Conduct

Please see [CODE OF CONDUCT](https://zaphyr.org/contributions#code-of-conduct) for details.

## License

This project is licensed under the MIT license. See [LICENSE](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file,
in reverse chronological order by release.

## [v2.0.1](https://github.com/zaphyr-org/encrypt/compare/2.0.0...2.0.1) [2023-11-19]

### New:
* Added `.vscode/` to .gitignore file

### Changed:
* Renamed `unit` to `phpunit` in composer.json scripts section
* Improved unit tests and moved tests to "Unit" directory

### Removed:
* Removed "tests" directory from phpstan.neon paths
* Removed phpstan-phpunit from composer require-dev

## [v2.0.0](https://github.com/zaphyr-org/encrypt/compare/1.0.1...2.0.0) [2023-04-05]

### New:
* First stable release version


tests/EncryptTest.php
<?php

declare(strict_types=1);

namespace Zaphyr\EncryptTests;

use PHPUnit\Framework\TestCase;
use Zaphyr\Encrypt\Encrypt;
use Zaphyr\Encrypt\Exceptions\DecryptException;
use Zaphyr\Encrypt\Exceptions\EncryptException;

class EncryptTest extends TestCase
{
    /**
     * ------------------------------------------
     * CONSTRUCTOR
     * ------------------------------------------
     */

    public function testConstructorThrowsExceptionOnLongerKey(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 32));
    }

    public function testConstructorThrowsExceptionWithShorterKeyLength(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 8));
    }

    public function testConstructorThrowsExceptionOnBadKeyLengthAndCustomCipher(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 16), 'AES-256-CBC');
    }

    public function testConstructorThrowsExceptionOnUnsupportedCipher(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 16), 'AES-256-CFB8');
    }

    /**
     * ------------------------------------------
     * ENCRYPT
     * ------------------------------------------
     */

    public function testEncrypt(): void
    {
        $encrypt = new Encrypt(str_repeat('a', 16));
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptString(): void
    {
        $encrypt = new Encrypt(str_repeat('a', 16));
        $encrypted = $encrypt->encryptString($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decryptString($encrypted));
    }

    public function testEncryptUsingBase64EncodedKey(): void
    {
        $encrypt = new Encrypt(random_bytes(16));
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptWithCustomCipher(): void
    {
        $encrypt = new Encrypt(str_repeat('b', 32), 'AES-256-CBC');
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptWithCustomCipherAndBase64EncodedKey(): void
    {
        $encrypt = new Encrypt(random_bytes(32), 'AES-256-CBC');
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    /**
     * ------------------------------------------
     * DECRYPT
     * ------------------------------------------
     */

    public function testDecryptThrowsExceptionOnInvalidPayload(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt = new Encrypt(str_repeat('a', 16));
        $payload = $encrypt->encrypt('foo');
        $payload = str_shuffle($payload);

        $encrypt->decrypt($payload);
    }

    public function testDecryptThrowsExceptionOnDifferentKey(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt1 = new Encrypt(str_repeat('a', 16));
        $encrypt2 = new Encrypt(str_repeat('b', 16));

        $encrypt2->decrypt($encrypt1->encrypt('foo'));
    }

    public function testDecryptThrowsExceptionWhenIvIsTooLong(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt = new Encrypt(str_repeat('a', 16));
        $payload = $encrypt->encrypt('foo');
        $data = json_decode(base64_decode($payload), true);
        $data['iv'] .= $data['value'][0];
        $data['value'] = substr($data['value'], 1);
        $modifiedPayload = base64_encode(json_encode($data) ?: '');

        $encrypt->decrypt($modifiedPayload);
    }
}


tests/Unit/EncryptTest.php
<?php

declare(strict_types=1);

namespace Zaphyr\EncryptTests\Unit;

use PHPUnit\Framework\TestCase;
use Zaphyr\Encrypt\Encrypt;
use Zaphyr\Encrypt\Exceptions\DecryptException;
use Zaphyr\Encrypt\Exceptions\EncryptException;

class EncryptTest extends TestCase
{
    /**
     * ------------------------------------------
     * CONSTRUCTOR
     * ------------------------------------------
     */

    public function testConstructorThrowsExceptionOnLongerKey(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 32));
    }

    public function testConstructorThrowsExceptionWithShorterKeyLength(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 8));
    }

    public function testConstructorThrowsExceptionOnBadKeyLengthAndCustomCipher(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 16), 'AES-256-CBC');
    }

    public function testConstructorThrowsExceptionOnUnsupportedCipher(): void
    {
        $this->expectException(EncryptException::class);

        new Encrypt(str_repeat('a', 16), 'AES-256-CFB8');
    }

    /**
     * ------------------------------------------
     * ENCRYPT
     * ------------------------------------------
     */

    public function testEncrypt(): void
    {
        $encrypt = new Encrypt(str_repeat('a', 16));
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptString(): void
    {
        $encrypt = new Encrypt(str_repeat('a', 16));
        $encrypted = $encrypt->encryptString($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decryptString($encrypted));
    }

    public function testEncryptUsingBase64EncodedKey(): void
    {
        $encrypt = new Encrypt(random_bytes(16));
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptWithCustomCipher(): void
    {
        $encrypt = new Encrypt(str_repeat('b', 32), 'AES-256-CBC');
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    public function testEncryptWithCustomCipherAndBase64EncodedKey(): void
    {
        $encrypt = new Encrypt(random_bytes(32), 'AES-256-CBC');
        $encrypted = $encrypt->encrypt($expected = 'foo');

        self::assertNotSame($expected, $encrypted);
        self::assertSame($expected, $encrypt->decrypt($encrypted));
    }

    /**
     * ------------------------------------------
     * DECRYPT
     * ------------------------------------------
     */

    public function testDecryptThrowsExceptionOnInvalidPayload(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt = new Encrypt(str_repeat('a', 16));
        $payload = $encrypt->encrypt('foo');
        $payload = str_shuffle($payload);

        $encrypt->decrypt($payload);
    }

    public function testDecryptThrowsExceptionOnDifferentKey(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt1 = new Encrypt(str_repeat('a', 16));
        $encrypt2 = new Encrypt(str_repeat('b', 16));

        $encrypt2->decrypt($encrypt1->encrypt('foo'));
    }

    public function testDecryptThrowsExceptionWhenIvIsTooLong(): void
    {
        $this->expectException(DecryptException::class);

        $encrypt = new Encrypt(str_repeat('a', 16));
        $payload = $encrypt->encrypt('foo');
        $data = json_decode(base64_decode($payload), true);
        $data['iv'] .= $data['value'][0];
        $data['value'] = substr($data['value'], 1);
        $modifiedPayload = base64_encode(json_encode($data) ?: '');

        $encrypt->decrypt($modifiedPayload);
    }
}


src/Contracts/EncryptInterface.php
<?php

declare(strict_types=1);

namespace Zaphyr\Encrypt\Contracts;

use Zaphyr\Encrypt\Exceptions\DecryptException;
use Zaphyr\Encrypt\Exceptions\EncryptException;

/**
 * @author merloxx <merloxx@zaphyr.org>
 */
interface EncryptInterface
{
    /**
     * @param mixed $value
     * @param bool  $serialize
     *
     * @throws EncryptException If the data could not be encrypted.
     * @return string
     */
    public function encrypt(mixed $value, bool $serialize = true): string;

    /**
     * @param string $value
     *
     * @throws EncryptException If the data could not be encrypted.
     * @return string
     */
    public function encryptString(string $value): string;

    /**
     * @param mixed $payload
     * @param bool  $unserialize
     *
     * @throws DecryptException If the data could not be decrypted.
     * @return mixed
     */
    public function decrypt(mixed $payload, bool $unserialize = true): mixed;

    /**
     * @param string $payload
     *
     * @throws DecryptException If the data could not be decrypted.
     * @return string
     */
    public function decryptString(string $payload): string;
}


src/Encrypt.php
<?php

declare(strict_types=1);

namespace Zaphyr\Encrypt;

use Zaphyr\Encrypt\Contracts\EncryptInterface;
use Zaphyr\Encrypt\Exceptions\DecryptException;
use Zaphyr\Encrypt\Exceptions\EncryptException;

/**
 * @author merloxx <merloxx@zaphyr.org>
 */
class Encrypt implements EncryptInterface
{
    /**
     * @var string
     */
    protected string $key;

    /**
     * @var string
     */
    protected string $cipher;

    /**
     * @param string $key
     * @param string $cipher
     *
     * @throws EncryptException
     */
    public function __construct(string $key, string $cipher = 'AES-128-CBC')
    {
        if (!$this->validKeyAndCipher($key, $cipher)) {
            throw new EncryptException(
                'The only supported ciphers are "AES-128-CBC" and "AES-256-CBC" with the correct key lengths'
            );
        }

        $this->key = $key;
        $this->cipher = $cipher;
    }

    /**
     * {@inheritdoc}
     */
    public function encrypt(mixed $value, bool $serialize = true): string
    {
        $iv = random_bytes(openssl_cipher_iv_length($this->cipher));
        $value = openssl_encrypt($serialize ? serialize($value) : $value, $this->cipher, $this->key, 0, $iv);

        if (!$value) {
            throw new EncryptException('Could not encrypt the data');
        }

        $mac = $this->hash($iv = base64_encode($iv), $value);
        $json = json_encode(compact('iv', 'value', 'mac'));

        if (!is_string($json) || json_last_error() !== JSON_ERROR_NONE) {
            throw new EncryptException('Could not encrypt the data');
        }

        return base64_encode($json);
    }

    /**
     * {@inheritdoc}
     */
    public function encryptString(string $value): string
    {
        return $this->encrypt($value, false);
    }

    /**
     * {@inheritdoc}
     */
    public function decrypt(mixed $payload, bool $unserialize = true): mixed
    {
        $payload = $this->getJsonPayload($payload);
        $iv = base64_decode($payload['iv']);
        $decrypted = openssl_decrypt($payload['value'], $this->cipher, $this->key, 0, $iv);

        if (!$decrypted) {
            throw new DecryptException('Could not decrypt data');
        }

        return $unserialize ? unserialize($decrypted) : $decrypted;
    }

    /**
     * {@inheritdoc}
     */
    public function decryptString(string $payload): string
    {
        return $this->decrypt($payload, false);
    }

    /**
     * @param string $key
     * @param string $cipher
     *
     * @return bool
     */
    protected function validKeyAndCipher(string $key, string $cipher): bool
    {
        $length = mb_strlen($key, '8bit');

        return ($cipher === 'AES-128-CBC' && $length === 16) || ($cipher === 'AES-256-CBC' && $length === 32);
    }

    /**
     * @param string $iv
     * @param mixed  $value
     *
     * @return string
     */
    protected function hash(string $iv, mixed $value): string
    {
        return hash_hmac('sha256', $iv . $value, $this->key);
    }

    /**
     * @param mixed $payload
     *
     * @throws DecryptException
     * @return array<string, mixed>
     */
    protected function getJsonPayload(mixed $payload): array
    {
        $payload = json_decode(base64_decode($payload), true);

        if (!$this->validPayload($payload)) {
            throw new DecryptException('Invalid payload');
        }

        if (!$this->validMac($payload)) {
            throw new DecryptException('Invalid MAC');
        }

        return $payload;
    }

    /**
     * @param mixed $payload
     *
     * @return bool
     */
    protected function validPayload(mixed $payload): bool
    {
        return is_array($payload) && isset($payload['iv'], $payload['value'], $payload['mac']) &&
            strlen((string)base64_decode($payload['iv'], true)) === openssl_cipher_iv_length($this->cipher);
    }

    /**
     * @param array<string, mixed> $payload
     *
     * @return bool
     */
    protected function validMac(array $payload): bool
    {
        return hash_equals($this->hash($payload['iv'], $payload['value']), $payload['mac']);
    }
}
